controllers/cloud.redhat.com/clowdapp_controller.go

```diff
@@ -433,6 +441,67 @@ func (r *ClowdAppReconciler) appsToEnqueueUponEnvUpdate(a client.Object) []recon
 
 	return reqs
 }
+func (r *ClowdAppReconciler) appToEnqueueOnJobCompletion(a client.Object) []reconcile.Request {
+	reqs := []reconcile.Request{}
+	ctx := context.Background()
+	obj := types.NamespacedName{
+		Name:      a.GetName(),
+		Namespace: a.GetNamespace(),
+	}
+
+	// Get the Job Resource
+	job := batchv1.Job{}
+	err := r.Client.Get(ctx, obj, &job)
+
+	if err != nil {
+		if k8serr.IsNotFound(err) {
+			// Must have been deleted
+			return reqs
+		}
+		r.Log.Error(err, "Failed to fetch Job")
+		return nil
+	}
+
+	switch {
+	case job.Status.Active == 1:
+		return reqs
+	case job.Status.Failed == 1:
+		return reqs
+	case job.Status.Succeeded == 0:
+		return reqs
+	}
+
+	// Get all the ClowdApp resources
+
+	appList := crd.ClowdAppList{}
+	err = r.Client.List(ctx, &appList)
+	if err != nil {
+		r.Log.Error(err, "Failed to list ClowdApps")
+		return nil
+	}
+
+	// Filter based on job name
+	for _, app := range appList.Items {
+		// find the app that the job belongs to
+		if job.OwnerReferences[0].Name == app.Name {
+			// update the job with "done" state and mark the generation we last
+			// ran the job for.
+			app.ObjectMeta.Annotations["clowder/pre-hook-status"] = "done"
+			app.ObjectMeta.Annotations["clowder/pre-hook-generation"] = strconv.Itoa(int(app.GetGeneration()))
```

Comment:
Depending on when this gets triggered, it is perhaps a little bit of an assumption to say that we know for sure the job was run against this particular generation. If the spec has changed in the meantime, this may not be true.

Reply:
If however we encode the apps generation into the job when it is started, we know that _that_ generation of the the app was used to build and trigger the job initially.